nav2_mppi_controller/src/controller.cpp

```diff
@@ -98,6 +102,10 @@ geometry_msgs::msg::TwistStamped MPPIController::computeVelocityCommands(
   RCLCPP_INFO(logger_, "Control loop execution time: %ld [ms]", duration);
 #endif
 
+  if (optimal_path_pub_->get_subscription_count() > 0) {
+    publish_optimal_path(optimizer_.getOptimizedTrajectory());
```

Comment:
`getOptimizedTrajectory` involves populating the return each call https://github.com/ros-navigation/navigation2/blob/1c56c7a532a8df9783c742961e341542e1e609e6/nav2_mppi_controller/src/optimizer.cpp#L369-L384. Instead, you should include this publisher into the `trajectory visualizer` class, and publish like you do here, in the `visualize()` member so that (a) you call this function once and (b) so that all the visualization work is in its namesake class